presto-main/src/main/java/com/facebook/presto/operator/scalar/RegexpFunctions.java

```diff
@@ -78,17 +85,118 @@ public static Slice regexpReplace(@SqlType(StandardTypes.VARCHAR) Slice source,
     @ScalarFunction
     @SqlType(StandardTypes.VARCHAR)
-    public static Slice regexpReplace(@SqlType(StandardTypes.VARCHAR) Slice source, @SqlType(RegexpType.NAME) Pattern pattern, @SqlType(StandardTypes.VARCHAR) Slice replacement)
+    public static Slice regexpReplace(@SqlType(StandardTypes.VARCHAR) Slice source, @SqlType(RegexpType.NAME) Regex pattern, @SqlType(StandardTypes.VARCHAR) Slice replacement)
     {
-        Matcher matcher = pattern.matcher(source.toString(UTF_8));
-        String replaced = matcher.replaceAll(replacement.toString(UTF_8));
-        return Slices.copiedBuffer(replaced, UTF_8);
+        Matcher matcher = pattern.matcher(source.getBytes());
+        SliceOutput sliceOutput = new DynamicSliceOutput(source.length() + replacement.length() * 5);
+
+        int lastEnd = 0;
+        int nextStart = 0;
+        while (true) {
+            int offset = matcher.search(nextStart, source.length(), Option.DEFAULT);
+            if (offset == -1) {
+                break;
+            }
+            if (matcher.getEnd() == matcher.getBegin()) {
+                nextStart = matcher.getEnd() + 1;
+            }
+            else {
+                nextStart = matcher.getEnd();
+            }
+            Slice sliceBetweenReplacements = source.slice(lastEnd, matcher.getBegin() - lastEnd);
+            lastEnd = matcher.getEnd();
+            sliceOutput.appendBytes(sliceBetweenReplacements);
+            appendReplacement(sliceOutput, source, pattern, matcher.getEagerRegion(), replacement);
+        }
+        sliceOutput.appendBytes(source.slice(lastEnd, source.length() - lastEnd));
+
+        return sliceOutput.slice();
+    }
+
+    private static void appendReplacement(SliceOutput result, Slice source, Regex pattern, Region region, Slice replacement)
+    {
+        // Handle the following items:
+        // 1. ${name};
+        // 2. $0, $1, $123 (group 123, if exists; or group 12, if exists; or group 1);
+        // 3. \\, \$, \t (literal 't').
+        // 4. Anything that doesn't starts with \ or $ is considered regular bytes
+
+        int idx = 0;
+
+        while (idx < replacement.length()) {
+            byte nextByte = replacement.getByte(idx);
+            if (nextByte == '$') {
+                idx++;
+                if (idx == replacement.length()) { // not using checkArgument because `.toStringUtf8` is expensive
+                    throw new PrestoException(INVALID_FUNCTION_ARGUMENT, "Illegal replacement sequence: " + replacement.toStringUtf8());
+                }
+                nextByte = replacement.getByte(idx);
+                int backref;
+                if (nextByte == '{') { // case 1 in the above comment
+                    idx++;
+                    int startCursor = idx;
+                    while (idx < replacement.length()) {
+                        nextByte = replacement.getByte(idx);
+                        if (nextByte == '}') {
```

Comment:
If group name is unicode could this fail? Can the '}' byte be inside another codepoint?